### PR TITLE
fix(custom-resources): provider framework will always log all data including confidential data

### LINKED ISSUE #30275: custom-resources: Provider logs Data from response with NoEcho: true
### Describe the bug

When using a Provider to create a custom resource, the request and response objects are logged by the provider function. There is no apparent way to prevent or redact this logging, resulting in secrets being logged if returned in the custom resource's Data object. By extension, if secret values are passed in the resource's ResourceProperties they will be logged as well.

### Expected Behavior

When the custom resource response has `NoEcho: true`, the log output from the Provider function should redact the values from the Data object.
```
[provider-framework] onEvent returned: 
{
    "NoEcho": true,
    "PhysicalResourceId": "2262225",
    "Data": {
        "clientId": "***",
        "clientSecret": "***"
    },
    "Status": "SUCCESS"
}
```

### Current Behavior

The provider function logged the full Data payload
```
[provider-framework] onEvent returned: 
{
    "NoEcho": true,
    "PhysicalResourceId": "2262225",
    "Data": {
        "clientId": "3a415657c61047fe9b790501254",
        "clientSecret": "475343b8<manually redacted>"
    },
    "Status": "SUCCESS"
}
```

### Reproduction Steps

```ts
import { App, Stack } from 'aws-cdk-lib';
import { Provider } from 'aws-cdk-lib/custom-resources';
import { Code, Function, Runtime } from 'aws-cdk-lib/aws-lambda';

const app = new App();
const stack = new Stack(app, 'cr-demo-stack');

const handler = new Function(stack , 'my-handler', {
  runtime: Runtime.NODEJS_20_X,
  handler: 'index.handler',
  code: Code.fromInline(`
  exports.handler = async (event, context) => {
    return {
      PhysicalResourceId: '1234',
      NoEcho: true,
      Data: {
        mySecret: 'secret-value',
      },
    };
  };`),
});

const provider = new Provider(stack , 'my-provider', {
  onEventHandler: handler,
});

new CustomResource(stack , 'my-cr', {
  serviceToken: provider.serviceToken,
});
```

Deploy this stack and you can see the following log:
```
[provider-framework] event: {
  "PhysicalResourceId": "1234",
  "NoEcho": true,
  "Data": {
    "mySecret": "secret-value"
  }
}
[provider-framework] submit response to cloudformation <stack-id> {
  "Status": "SUCCESS",
  "Reason": "SUCCESS",
  "StackId": "<stack-id>",
  "RequestId": "bab8ac9b-c6a7-45d4-9828-71dc260ebef7",
  "PhysicalResourceId": "1234",
  "LogicalResourceId": "clientapplication",
  "NoEcho": true,
  "Data": {
    "mySecret": "secret-value"
  }
}
```

### Possible Solution

Add logic to the provider handler code to redact the Data object if NoEcho = true

Add properties to the Provider construct to redact some/all of the ResourceProperties from the provider logs.

### Additional Information/Context

_No response_

### CDK CLI Version

2.133.0 (build dcc1e75)

### Framework Version

2.133.0

### Node.js Version

20

### OS

Ubuntu

### Language

TypeScript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by GavinZZ**
### Issue # (if applicable)

Closes #30275.

### Reason for this change

When using a Provider to create a custom resource, the request and response objects are logged by the provider function. There is no apparent way to prevent or redact this logging, resulting in secrets being logged if returned in the custom resource's Data object. By extension, if secret values are passed in the resource's ResourceProperties they will be logged as well.

### Description of changes

Allow `NoEcho` fields to mask the data response to `*****`.

### Description of how you validated changes

Integ test covering this and verifeid in the log stream that `redacted` is included in the message.

### Checklist
- [x] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/custom-resources/README.md [hunk 1]
```diff
@@ -215,6 +215,48 @@ must return this name in `PhysicalResourceId` and make sure to handle
 replacement properly. The `S3File` example demonstrates this
 through the `objectKey` property.
 
+### Masking the output of log statements
+
+When using the Provider framework to create a custom resource, the request and response
+objects are logged by the provider function.If secret values are returned in the custom
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/custom-resources/README.md:221]:
```suggestion
objects are logged by the provider function. If secret values are returned in the custom
```

File: packages/aws-cdk-lib/custom-resources/lib/provider-framework/runtime/cfn-response.ts [hunk 2]
```diff
@@ -101,4 +106,18 @@ export function safeHandler(block: (event: any) => Promise<void>) {
...
 
+export function redactDataFromPayload(payload: OnEventResponse) {
+  // Create a deep copy of the payload object
+  const redactedPayload: OnEventResponse = JSON.parse(JSON.stringify(payload));
+
+  // Redact the data in the copied payload object
+  if (redactedPayload.Data) {
+    const keys = Object.keys(redactedPayload.Data);
+    for (const key of keys) {
+      redactedPayload.Data[key] = '*****';
```
Comment (paulhcsun) [inline packages/aws-cdk-lib/custom-resources/lib/provider-framework/runtime/cfn-response.ts:117]: nit:
```suggestion
      redactedPayload.Data.get(key) = '*****';
```

**Review by paulhcsun — COMMENTED** [review]
Minor changes, otherwise this looks good to me.

File: packages/aws-cdk-lib/custom-resources/lib/provider-framework/runtime/cfn-response.ts (hunk 2 above)
Comment (GavinZZ) [reply packages/aws-cdk-lib/custom-resources/lib/provider-framework/runtime/cfn-response.ts:117]: I think that's python syntax? Don't think typescript supports `{}.get()` call.

**Review by paulhcsun — APPROVED** [review]
noice @mergifyio update

File: packages/aws-cdk-lib/custom-resources/lib/provider-framework/runtime/cfn-response.ts (hunk 2 above)
Comment (paulhcsun) [reply packages/aws-cdk-lib/custom-resources/lib/provider-framework/runtime/cfn-response.ts:117]: u right lol

**Pull request merged**